src/ralph/schemas/edx/browser.py

```diff
@@ -0,0 +1,330 @@
+                f"Event new ({event['new']}) should be equal to old "
+                f"({event['old']}) + diff ({diff})"
+            )
+
+    @validates_schema
+    def validate_event_textbook_pdf(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 3 keys:
+        `name`: string, `page`: integer > 0, `chapter`: url
+        and `name` should be equal to event `name`
+        """
+        if data["name"] not in (
+            "textbook.pdf.outline.toggled",
+            "textbook.pdf.page.navigated",
+            "textbook.pdf.thumbnails.toggled",
+        ):
+            return
+
+        event = self.validate_event_keys(data, {"chapter", "page", "name"})
+        self.check_chapter_page_name(data, event)
+
+    @validates_schema
+    def validate_event_textbook_pdf_thumbnail_navigated(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 4 keys:
+         `name`: string, `page`: integer > 0, `chapter`: url
+         `thumbnail_title`: string """
+        if data["name"] != "textbook.pdf.thumbnail.navigated":
+            return
+
+        event = self.validate_event_keys(
+            data, {"chapter", "page", "name", "thumbnail_title"}
+        )
+        self.check_chapter_page_name(data, event)
+        if not isinstance(event["thumbnail_title"], str):
+            raise ValidationError("thumbnail_title should be a string")
+
+    @validates_schema
+    def validate_event_textbook_pdf_zoom_buttons_changed(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 4 keys:
+        `name`: string, `page`: integer > 0, `chapter`: url,
+        `direction`: string (one of [`in`, `out`] or [`up`, `down`])
+        """
+        if data["name"] not in [
+            "textbook.pdf.zoom.buttons.changed",
+            "textbook.pdf.page.scrolled",
+        ]:
+            return
+
+        event = self.validate_event_keys(data, {"chapter", "page", "name", "direction"})
+        self.check_chapter_page_name(data, event)
+        valid_directions = (
+            ["in", "out"] if data["name"][-7:] == "changed" else ["up", "down"]
+        )
+        if event["direction"] not in valid_directions:
+            raise ValidationError(f"direction should be one of {valid_directions}")
+
+    @validates_schema
+    def validate_event_textbook_pdf_zoom_menu_changed(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 4 keys:
+        `name`: string, `page`: integer > 0, `chapter`: url,
+        `amaunt`: string
+        """
+        if data["name"] != "textbook.pdf.zoom.menu.changed":
+            return
+
+        event = self.validate_event_keys(data, {"chapter", "page", "name", "amaunt"})
+        self.check_chapter_page_name(data, event)
+        if event["amaunt"] not in BROWSER_EVENT_VALID_AMAUNT:
+            raise ValidationError(
+                f"amaunt should be one of {BROWSER_EVENT_VALID_AMAUNT}"
+            )
```

Comment:
IMO this is not a good idea to have a single `BrowserEventSchema` with x schema validation methods. This is not efficient at all! I would prefer having 1 schema per event and a `BaseBrowserEventSchema` to inherit from.

Reply:
Good Idea, I will give it a try, thanks